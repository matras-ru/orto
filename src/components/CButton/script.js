import { selfInstall, noop } from '@/utils';
import commonAttributes from '@/mixins/commonAttributes.js';

import ThemeClass from '@/themes/default/CButton';

const {
    baseClass,
    defaultClass,
    primaryClass,
    secondaryClass,
    tertiaryClass,
    quaternaryClass,
    disabledClass,
    defaultSizeClass,
    largeSizeClass,
    smallSizeClass
} = ThemeClass;

const validTagNames = ['button', 'a'];

const validVariants = ['primary', 'secondary', 'tertiary', 'quaternary'];

const validSizes = ['lg', 'sm'];

const props = {
    ...commonAttributes.props,
    tagName: {
        type: String,
        default: 'button',
        validator: value => validTagNames.indexOf(value) !== -1
    },
    label: {
        type: String,
        default: null
    },
    value: {
        type: [String, Number],
        default: null
    },
    type: {
        type: String,
        default: null
    },
    href: {
        type: String,
        default: null
    },
    variant: {
        type: String,
        default: null,
        validator: value => value === null || validVariants.indexOf(value) !== -1
    },
    size: {
        type: String,
        default: null,
        validator: value => value === null || validSizes.indexOf(value) !== -1
    },
    activeClass: {
        type: String,
        default: 'router-link-active'
    },
    exactActiveClass: {
        type: String,
        default: 'router-link-exact-active'
    }
};
/**
 * @description
 *
 * @return {Array}
 */
const currentClass = props => {
    let classes = [baseClass];

    if (props.disabled) {
        classes.push(disabledClass);
    }

    switch (props.size) {
        case 'sm':
            classes.push(smallSizeClass);
            break;
        case 'lg':
            classes.push(largeSizeClass);
            break;
        default:
            classes.push(defaultSizeClass);
            break;
    }

    switch (props.variant) {
        case 'primary':
            classes.push(primaryClass);
            break;
        case 'secondary':
            classes.push(secondaryClass);
            break;
        case 'tertiary':
            classes.push(tertiaryClass);
            break;
        case 'quaternary':
            classes.push(quaternaryClass);
            break;
        default:
            classes.push(defaultClass);
            break;
    }

    return classes;
};

const componentToRender = props => {
    // if (this.isARouterLink) {
    //     return (
    //         this.$options.components.NuxtLink ||
    //         this.$options.components.RouterLink
    //     );
    // }

    if (props.href) {
        return 'a';
    }

    return props.tagName;
};

/**
 *
 * @return {Object}
 */

const getAttributes = props => {
    // if (props.isARouterLink) {
    //     return {
    //         to: props.to,
    //         replace: props.replace,
    //         append: props.append,
    //         tag: props.tagName,
    //         activeClass: props.activeClass,
    //         exact: props.exact,
    //         event: ['click', 'focus', 'blur'],
    //         exactActiveClass: props.exactActiveClass,
    //         id: props.id,
    //         value: props.value,
    //         autofocus: props.autofocus,
    //         disabled: props.disabled,
    //         name: props.name,
    //         type: props.type
    //     };
    // }

    return {
        id: props.id,
        value: props.value,
        autofocus: props.autofocus,
        disabled: props.disabled,
        name: props.name,
        href: props.href,
        type: props.type
    };
};

export default {
    name: 'Button',

    functional: true,

    install(Vue, theme) {
        selfInstall(Vue, theme, this);
    },

    props,

    render(h, { props, listeners, children }) {
        const onClick = listeners['onClick'] || noop;
        const onFocus = listeners['onFocus'] || noop;
        const onBlur = listeners['onBlur'] || noop;

        const on = {
            click(e) {
                onClick(e);
            },

            focus(e) {
                onFocus(e);
            },

            blur(e) {
                onBlur(e);
            }
        };

        const componentData = {
            class: currentClass(props),
            attrs: getAttributes(props),
            on
        };

        return h(componentToRender(props), componentData, props.label ? props.label : children);
    }
};
