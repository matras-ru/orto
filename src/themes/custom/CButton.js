export const baseClass =
    'inline-block align-top rounded-lg uppercase font-semibold text-black-100 duration-250 ease-in-out border-3 mb-1-4';
export const defaultClass = 'bg-white border-primary-100 transition-shadow hover:shadow';
export const primaryClass = 'bg-primary-100 border-primary-100 transition-shadow hover:shadow';
export const secondaryClass =
    'bg-white border-secondary-200 text-secondary-200 transition-colors hover:text-white hover:bg-secondary-200';
export const tertiaryClass =
    'bg-white border-tertiary-100 text-secondary-200 transition-colors hover:bg-tertiary-100';
export const quaternaryClass =
    'bg-white border-black-100 transition-colors hover:text-white hover:bg-black-100';
export const disabledClass = 'cursor-not-allowed opacity-75';
export const defaultSizeClass = 'text-base px-1-5 py-0-4 leading-snug';
export const largeSizeClass = 'text-lg px-1-5 py-0-6 leading-none';
export const smallSizeClass = 'text-base px-1-3 py-0-3 leading-snug';

export default {
    baseClass,
    defaultClass,
    primaryClass,
    secondaryClass,
    tertiaryClass,
    quaternaryClass,
    disabledClass,
    defaultSizeClass,
    smallSizeClass,
    largeSizeClass
};
