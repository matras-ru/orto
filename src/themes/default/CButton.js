export const base =
    'inline-flex items-center justify-center rounded-lg uppercase font-semibold text-black-100 duration-250 ease-in-out border-3 focus:outline-none';

export const variantPrimary =
    'border-primary-100 transition-shadow hover:shadow disabled:shadow-none';
export const variantSecondary =
    'bg-primary-100 border-primary-100 transition-shadow hover:shadow disabled:shadow-none';
export const variantTertiary =
    'border-secondary-200 text-secondary-200 transition-colors hover:text-white disabled:text-secondary-200 hover:bg-secondary-200 disabled:bg-transparent';
export const variantQuaternary =
    'border-tertiary-100 text-secondary-200 transition-bg hover:bg-tertiary-100 disabled:bg-transparent';
export const variantQuinary =
    'border-black-100 transition-colors hover:text-white disabled:text-black-100 hover:bg-black-100 disabled:bg-transparent';

export const stateDisable = 'cursor-not-allowed opacity-75';

export const sizeSm = 'text-base px-1-3 py-0-3 leading-snug';
export const sizeMd = 'text-base px-1-5 py-0-4 leading-snug';
export const sizeLg = 'text-lg px-1-5 py-0-6 leading-none';

export const displayBlock = 'w-full';

export default {
    base,

    variantPrimary,
    variantSecondary,
    variantTertiary,
    variantQuaternary,
    variantQuinary,

    stateDisable,

    sizeSm,
    sizeMd,
    sizeLg,

    displayBlock
};
