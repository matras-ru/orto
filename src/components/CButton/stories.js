import { storiesOf } from '@storybook/vue';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';

import CButton from './';

const stories = storiesOf('CButton', module);
stories.addDecorator(withKnobs);

stories
    .add('Button knobs', () => ({
        components: { CButton },
        props: {
            id: {
                default: text('id', 'id')
            },
            value: {
                default: text('value (string, number)', 'value')
            },
            type: {
                default: text('type (button, reset, submit)', 'button')
            },
            size: {
                default: text('size (sm, lg)', '')
            },
            variant: {
                default: text('variant (primary, secondary, tertiary, quaternary)', '')
            },
            autofocus: {
                default: boolean('autofocus', false)
            },
            disabled: {
                default: boolean('disabled', false)
            }
        },
        template:
            '<CButton @onClick="action" :id="id" :value="value" :tagName="\'button\'" :type="type" :size="size" :variant="variant" :autofocus="autofocus" :disabled="disabled">Button</CButton>',
        methods: { action: action('clicked') }
    }))
    .add('Link knobs', () => ({
        components: { CButton },
        props: {
            id: {
                default: text('id', 'id')
            },
            href: {
                default: text('href', '/')
            },
            size: {
                default: text('size (sm, lg)', '')
            },
            variant: {
                default: text('variant (primary, secondary, tertiary, quaternary)', '')
            },
            autofocus: {
                default: boolean('autofocus', false)
            },
            disabled: {
                default: boolean('disabled', false)
            }
        },
        template:
            '<CButton @onClick="action" :id="id" :tagName="\'a\'" :href="href" :size="size" :variant="variant" :autofocus="autofocus" :disabled="disabled">Link</CButton>',
        methods: { action: action('clicked') }
    }))
    .add('Default', () => ({
        components: { CButton },
        template:
            '<div><CButton @onClick="action" size="sm">Button</CButton><CButton @onClick="action">Button</CButton><CButton @onClick="action" size="lg">Button</CButton></div>',
        methods: { action: action('clicked') }
    }))
    .add('Primary', () => ({
        components: { CButton },
        template:
            '<div><CButton @onClick="action" variant="primary" size="sm">Button</CButton><CButton @onClick="action" variant="primary">Button</CButton><CButton @onClick="action" variant="primary" size="lg">Button</CButton></div>',
        methods: { action: action('clicked') }
    }))
    .add('Secondary', () => ({
        components: { CButton },
        template:
            '<div><CButton @onClick="action" variant="secondary" size="sm">Button</CButton><CButton @onClick="action" variant="secondary">Button</CButton><CButton @onClick="action" variant="secondary" size="lg">Button</CButton></div>',
        methods: { action: action('clicked') }
    }))
    .add('Tertiary', () => ({
        components: { CButton },
        template:
            '<div><CButton @onClick="action" variant="tertiary" size="sm">Button</CButton><CButton @onClick="action" variant="tertiary">Button</CButton><CButton @onClick="action" variant="tertiary" size="lg">Button</CButton></div>',
        methods: { action: action('clicked') }
    }))
    .add('Quaternary', () => ({
        components: { CButton },
        template:
            '<div><CButton @onClick="action" variant="quaternary" size="sm">Button</CButton><CButton @onClick="action" variant="quaternary">Button</CButton><CButton @onClick="action" variant="quaternary" size="lg">Button</CButton></div>',
        methods: { action: action('clicked') }
    }));
