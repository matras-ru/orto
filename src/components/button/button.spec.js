/* eslint-env jest */
import { mount } from '@vue/test-utils';
import CButton from './button';

const baseClass =
    'inline-block align-top rounded-lg uppercase font-semibold text-black-100 transition-250 transition-ease-in-out border-3';
const defaultClass = 'bg-white border-primary-100 transition-shadow hover:shadow';
const secondaryClass = 'bg-primary-100 border-primary-100 transition-shadow hover:shadow';
const disabledClass = 'cursor-not-allowed opacity-75';
const defaultSizeClass = 'text-base px-1-5 py-0-4 leading-snug';

const baseLinkClass = 'inline-block no-underline';
const variantPrimaryLink = 'text-secondary-200 hover:text-black-200 border-b-2';

describe('Button', () => {
    it('The button is rendered', () => {
        const wrapper = mount(CButton);

        expect(wrapper.is('button')).toBe(true);
        expect(wrapper.attributes('href')).not.toBeDefined();
        expect(wrapper.classes().sort()).toEqual(
            `${baseClass} ${defaultClass} ${defaultSizeClass}`.split(' ').sort()
        );
    });

    it('Applies variant class', () => {
        const wrapper = mount(CButton, {
            context: {
                props: {
                    variant: 'secondary'
                }
            }
        });

        expect(wrapper.is('button')).toBe(true);
        expect(wrapper.classes().sort()).toEqual(
            `${baseClass} ${secondaryClass} ${defaultSizeClass}`.split(' ').sort()
        );
    });

    it('Should emit click event when clicked', () => {
        let called = 0;
        let evt = null;

        const wrapper = mount(CButton, {
            listeners: {
                onClick: e => {
                    evt = e;
                    called++;
                }
            }
        });

        expect(wrapper.is('button')).toBe(true);
        expect(called).toBe(0);
        expect(evt).toEqual(null);
        wrapper.trigger('click');
        expect(called).toBe(1);
        expect(evt).toBeInstanceOf(MouseEvent);
    });

    it('The button is disabled', () => {
        const wrapper = mount(CButton, {
            context: {
                props: {
                    disabled: true
                }
            }
        });

        expect(wrapper.attributes('disabled')).toBeDefined();
        expect(wrapper.classes().sort()).toEqual(
            `${baseClass} ${defaultClass} ${defaultSizeClass} ${disabledClass}`.split(' ').sort()
        );
    });

    it('Should not emit click event when clicked and disabled', () => {
        let called = 0;
        let evt = null;

        const wrapper = mount(CButton, {
            context: {
                props: {
                    disabled: true
                },
                listeners: {
                    onClick: e => {
                        evt = e;
                        called++;
                    }
                }
            }
        });

        expect(called).toBe(0);
        expect(evt).toEqual(null);
        expect(wrapper.attributes('disabled')).toBeDefined();
        wrapper.trigger('click');
        expect(evt).toEqual(null);
        expect(called).toBe(0);
    });

    it('Button as link - missing link classes', () => {
        const wrapper = mount(CButton, {
            context: {
                props: {
                    href: '//google.com',
                    target: '_blank'
                }
            }
        });

        expect(wrapper.is('a')).toBe(true);
        expect(wrapper.classes('no-underline')).toBe(false);
        expect(wrapper.classes('text-secondary-200')).toBe(false);
        expect(wrapper.classes('hover:text-black-200')).toBe(false);
        expect(wrapper.classes('border-b-2')).toBe(false);
    });
});
