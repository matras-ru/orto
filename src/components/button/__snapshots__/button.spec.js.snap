// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Applies variant class 1`] = `<button class="inline-flex items-center justify-center rounded-lg uppercase font-semibold text-black-100 duration-250 ease-in-out border-3 focus:outline-none text-base px-1-5 py-0-4 leading-snug bg-primary-100 border-primary-100 transition-shadow hover:shadow disabled:shadow-none"></button>`;

exports[`Button Button as link - missing link classes 1`] = `<a href="//google.com" target="_blank" class="inline-flex items-center justify-center rounded-lg uppercase font-semibold text-black-100 duration-250 ease-in-out border-3 focus:outline-none text-base px-1-5 py-0-4 leading-snug border-primary-100 transition-shadow hover:shadow disabled:shadow-none"></a>`;

exports[`Button The button is disabled 1`] = `<button disabled="disabled" class="inline-flex items-center justify-center rounded-lg uppercase font-semibold text-black-100 duration-250 ease-in-out border-3 focus:outline-none text-base px-1-5 py-0-4 leading-snug border-primary-100 transition-shadow hover:shadow disabled:shadow-none cursor-not-allowed opacity-75"></button>`;

exports[`Button The button is rendered 1`] = `<button class="inline-flex items-center justify-center rounded-lg uppercase font-semibold text-black-100 duration-250 ease-in-out border-3 focus:outline-none text-base px-1-5 py-0-4 leading-snug border-primary-100 transition-shadow hover:shadow disabled:shadow-none"></button>`;
